// @zoralabs/nft-hooks/dist/graph-queries/ens-graph-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._SubgraphErrorPolicy_ = exports.Transfer_OrderBy = exports.TextChanged_OrderBy = exports.Resolver_OrderBy = exports.ResolverEvent_OrderBy = exports.Registration_OrderBy = exports.RegistrationEvent_OrderBy = exports.PubkeyChanged_OrderBy = exports.OrderDirection = exports.NewTtl_OrderBy = exports.NewResolver_OrderBy = exports.NewOwner_OrderBy = exports.NameTransferred_OrderBy = exports.NameRenewed_OrderBy = exports.NameRegistered_OrderBy = exports.NameChanged_OrderBy = exports.MulticoinAddrChanged_OrderBy = exports.InterfaceChanged_OrderBy = exports.Domain_OrderBy = exports.DomainEvent_OrderBy = exports.ContenthashChanged_OrderBy = exports.AuthorisationChanged_OrderBy = exports.AddrChanged_OrderBy = exports.Account_OrderBy = exports.AbiChanged_OrderBy = void 0;
var AbiChanged_OrderBy;
(function (AbiChanged_OrderBy) {
    AbiChanged_OrderBy["Id"] = "id";
    AbiChanged_OrderBy["Resolver"] = "resolver";
    AbiChanged_OrderBy["BlockNumber"] = "blockNumber";
    AbiChanged_OrderBy["TransactionId"] = "transactionID";
    AbiChanged_OrderBy["ContentType"] = "contentType";
})(AbiChanged_OrderBy = exports.AbiChanged_OrderBy || (exports.AbiChanged_OrderBy = {}));
var Account_OrderBy;
(function (Account_OrderBy) {
    Account_OrderBy["Id"] = "id";
    Account_OrderBy["Domains"] = "domains";
    Account_OrderBy["Registrations"] = "registrations";
})(Account_OrderBy = exports.Account_OrderBy || (exports.Account_OrderBy = {}));
var AddrChanged_OrderBy;
(function (AddrChanged_OrderBy) {
    AddrChanged_OrderBy["Id"] = "id";
    AddrChanged_OrderBy["Resolver"] = "resolver";
    AddrChanged_OrderBy["BlockNumber"] = "blockNumber";
    AddrChanged_OrderBy["TransactionId"] = "transactionID";
    AddrChanged_OrderBy["Addr"] = "addr";
})(AddrChanged_OrderBy = exports.AddrChanged_OrderBy || (exports.AddrChanged_OrderBy = {}));
var AuthorisationChanged_OrderBy;
(function (AuthorisationChanged_OrderBy) {
    AuthorisationChanged_OrderBy["Id"] = "id";
    AuthorisationChanged_OrderBy["Resolver"] = "resolver";
    AuthorisationChanged_OrderBy["BlockNumber"] = "blockNumber";
    AuthorisationChanged_OrderBy["TransactionId"] = "transactionID";
    AuthorisationChanged_OrderBy["Owner"] = "owner";
    AuthorisationChanged_OrderBy["Target"] = "target";
    AuthorisationChanged_OrderBy["IsAuthorized"] = "isAuthorized";
})(AuthorisationChanged_OrderBy = exports.AuthorisationChanged_OrderBy || (exports.AuthorisationChanged_OrderBy = {}));
var ContenthashChanged_OrderBy;
(function (ContenthashChanged_OrderBy) {
    ContenthashChanged_OrderBy["Id"] = "id";
    ContenthashChanged_OrderBy["Resolver"] = "resolver";
    ContenthashChanged_OrderBy["BlockNumber"] = "blockNumber";
    ContenthashChanged_OrderBy["TransactionId"] = "transactionID";
    ContenthashChanged_OrderBy["Hash"] = "hash";
})(ContenthashChanged_OrderBy = exports.ContenthashChanged_OrderBy || (exports.ContenthashChanged_OrderBy = {}));
var DomainEvent_OrderBy;
(function (DomainEvent_OrderBy) {
    DomainEvent_OrderBy["Id"] = "id";
    DomainEvent_OrderBy["Domain"] = "domain";
    DomainEvent_OrderBy["BlockNumber"] = "blockNumber";
    DomainEvent_OrderBy["TransactionId"] = "transactionID";
})(DomainEvent_OrderBy = exports.DomainEvent_OrderBy || (exports.DomainEvent_OrderBy = {}));
var Domain_OrderBy;
(function (Domain_OrderBy) {
    Domain_OrderBy["Id"] = "id";
    Domain_OrderBy["Name"] = "name";
    Domain_OrderBy["LabelName"] = "labelName";
    Domain_OrderBy["Labelhash"] = "labelhash";
    Domain_OrderBy["Parent"] = "parent";
    Domain_OrderBy["Subdomains"] = "subdomains";
    Domain_OrderBy["ResolvedAddress"] = "resolvedAddress";
    Domain_OrderBy["Owner"] = "owner";
    Domain_OrderBy["Resolver"] = "resolver";
    Domain_OrderBy["Ttl"] = "ttl";
    Domain_OrderBy["IsMigrated"] = "isMigrated";
    Domain_OrderBy["Events"] = "events";
})(Domain_OrderBy = exports.Domain_OrderBy || (exports.Domain_OrderBy = {}));
var InterfaceChanged_OrderBy;
(function (InterfaceChanged_OrderBy) {
    InterfaceChanged_OrderBy["Id"] = "id";
    InterfaceChanged_OrderBy["Resolver"] = "resolver";
    InterfaceChanged_OrderBy["BlockNumber"] = "blockNumber";
    InterfaceChanged_OrderBy["TransactionId"] = "transactionID";
    InterfaceChanged_OrderBy["InterfaceId"] = "interfaceID";
    InterfaceChanged_OrderBy["Implementer"] = "implementer";
})(InterfaceChanged_OrderBy = exports.InterfaceChanged_OrderBy || (exports.InterfaceChanged_OrderBy = {}));
var MulticoinAddrChanged_OrderBy;
(function (MulticoinAddrChanged_OrderBy) {
    MulticoinAddrChanged_OrderBy["Id"] = "id";
    MulticoinAddrChanged_OrderBy["Resolver"] = "resolver";
    MulticoinAddrChanged_OrderBy["BlockNumber"] = "blockNumber";
    MulticoinAddrChanged_OrderBy["TransactionId"] = "transactionID";
    MulticoinAddrChanged_OrderBy["CoinType"] = "coinType";
    MulticoinAddrChanged_OrderBy["Addr"] = "addr";
})(MulticoinAddrChanged_OrderBy = exports.MulticoinAddrChanged_OrderBy || (exports.MulticoinAddrChanged_OrderBy = {}));
var NameChanged_OrderBy;
(function (NameChanged_OrderBy) {
    NameChanged_OrderBy["Id"] = "id";
    NameChanged_OrderBy["Resolver"] = "resolver";
    NameChanged_OrderBy["BlockNumber"] = "blockNumber";
    NameChanged_OrderBy["TransactionId"] = "transactionID";
    NameChanged_OrderBy["Name"] = "name";
})(NameChanged_OrderBy = exports.NameChanged_OrderBy || (exports.NameChanged_OrderBy = {}));
var NameRegistered_OrderBy;
(function (NameRegistered_OrderBy) {
    NameRegistered_OrderBy["Id"] = "id";
    NameRegistered_OrderBy["Registration"] = "registration";
    NameRegistered_OrderBy["BlockNumber"] = "blockNumber";
    NameRegistered_OrderBy["TransactionId"] = "transactionID";
    NameRegistered_OrderBy["Registrant"] = "registrant";
    NameRegistered_OrderBy["ExpiryDate"] = "expiryDate";
})(NameRegistered_OrderBy = exports.NameRegistered_OrderBy || (exports.NameRegistered_OrderBy = {}));
var NameRenewed_OrderBy;
(function (NameRenewed_OrderBy) {
    NameRenewed_OrderBy["Id"] = "id";
    NameRenewed_OrderBy["Registration"] = "registration";
    NameRenewed_OrderBy["BlockNumber"] = "blockNumber";
    NameRenewed_OrderBy["TransactionId"] = "transactionID";
    NameRenewed_OrderBy["ExpiryDate"] = "expiryDate";
})(NameRenewed_OrderBy = exports.NameRenewed_OrderBy || (exports.NameRenewed_OrderBy = {}));
var NameTransferred_OrderBy;
(function (NameTransferred_OrderBy) {
    NameTransferred_OrderBy["Id"] = "id";
    NameTransferred_OrderBy["Registration"] = "registration";
    NameTransferred_OrderBy["BlockNumber"] = "blockNumber";
    NameTransferred_OrderBy["TransactionId"] = "transactionID";
    NameTransferred_OrderBy["NewOwner"] = "newOwner";
})(NameTransferred_OrderBy = exports.NameTransferred_OrderBy || (exports.NameTransferred_OrderBy = {}));
var NewOwner_OrderBy;
(function (NewOwner_OrderBy) {
    NewOwner_OrderBy["Id"] = "id";
    NewOwner_OrderBy["Domain"] = "domain";
    NewOwner_OrderBy["BlockNumber"] = "blockNumber";
    NewOwner_OrderBy["TransactionId"] = "transactionID";
    NewOwner_OrderBy["Owner"] = "owner";
})(NewOwner_OrderBy = exports.NewOwner_OrderBy || (exports.NewOwner_OrderBy = {}));
var NewResolver_OrderBy;
(function (NewResolver_OrderBy) {
    NewResolver_OrderBy["Id"] = "id";
    NewResolver_OrderBy["Domain"] = "domain";
    NewResolver_OrderBy["BlockNumber"] = "blockNumber";
    NewResolver_OrderBy["TransactionId"] = "transactionID";
    NewResolver_OrderBy["Resolver"] = "resolver";
})(NewResolver_OrderBy = exports.NewResolver_OrderBy || (exports.NewResolver_OrderBy = {}));
var NewTtl_OrderBy;
(function (NewTtl_OrderBy) {
    NewTtl_OrderBy["Id"] = "id";
    NewTtl_OrderBy["Domain"] = "domain";
    NewTtl_OrderBy["BlockNumber"] = "blockNumber";
    NewTtl_OrderBy["TransactionId"] = "transactionID";
    NewTtl_OrderBy["Ttl"] = "ttl";
})(NewTtl_OrderBy = exports.NewTtl_OrderBy || (exports.NewTtl_OrderBy = {}));
var OrderDirection;
(function (OrderDirection) {
    OrderDirection["Asc"] = "asc";
    OrderDirection["Desc"] = "desc";
})(OrderDirection = exports.OrderDirection || (exports.OrderDirection = {}));
var PubkeyChanged_OrderBy;
(function (PubkeyChanged_OrderBy) {
    PubkeyChanged_OrderBy["Id"] = "id";
    PubkeyChanged_OrderBy["Resolver"] = "resolver";
    PubkeyChanged_OrderBy["BlockNumber"] = "blockNumber";
    PubkeyChanged_OrderBy["TransactionId"] = "transactionID";
    PubkeyChanged_OrderBy["X"] = "x";
    PubkeyChanged_OrderBy["Y"] = "y";
})(PubkeyChanged_OrderBy = exports.PubkeyChanged_OrderBy || (exports.PubkeyChanged_OrderBy = {}));
var RegistrationEvent_OrderBy;
(function (RegistrationEvent_OrderBy) {
    RegistrationEvent_OrderBy["Id"] = "id";
    RegistrationEvent_OrderBy["Registration"] = "registration";
    RegistrationEvent_OrderBy["BlockNumber"] = "blockNumber";
    RegistrationEvent_OrderBy["TransactionId"] = "transactionID";
})(RegistrationEvent_OrderBy = exports.RegistrationEvent_OrderBy || (exports.RegistrationEvent_OrderBy = {}));
var Registration_OrderBy;
(function (Registration_OrderBy) {
    Registration_OrderBy["Id"] = "id";
    Registration_OrderBy["Domain"] = "domain";
    Registration_OrderBy["RegistrationDate"] = "registrationDate";
    Registration_OrderBy["ExpiryDate"] = "expiryDate";
    Registration_OrderBy["Registrant"] = "registrant";
    Registration_OrderBy["LabelName"] = "labelName";
    Registration_OrderBy["Events"] = "events";
})(Registration_OrderBy = exports.Registration_OrderBy || (exports.Registration_OrderBy = {}));
var ResolverEvent_OrderBy;
(function (ResolverEvent_OrderBy) {
    ResolverEvent_OrderBy["Id"] = "id";
    ResolverEvent_OrderBy["Resolver"] = "resolver";
    ResolverEvent_OrderBy["BlockNumber"] = "blockNumber";
    ResolverEvent_OrderBy["TransactionId"] = "transactionID";
})(ResolverEvent_OrderBy = exports.ResolverEvent_OrderBy || (exports.ResolverEvent_OrderBy = {}));
var Resolver_OrderBy;
(function (Resolver_OrderBy) {
    Resolver_OrderBy["Id"] = "id";
    Resolver_OrderBy["Domain"] = "domain";
    Resolver_OrderBy["Address"] = "address";
    Resolver_OrderBy["Addr"] = "addr";
    Resolver_OrderBy["ContentHash"] = "contentHash";
    Resolver_OrderBy["Texts"] = "texts";
    Resolver_OrderBy["CoinTypes"] = "coinTypes";
    Resolver_OrderBy["Events"] = "events";
})(Resolver_OrderBy = exports.Resolver_OrderBy || (exports.Resolver_OrderBy = {}));
var TextChanged_OrderBy;
(function (TextChanged_OrderBy) {
    TextChanged_OrderBy["Id"] = "id";
    TextChanged_OrderBy["Resolver"] = "resolver";
    TextChanged_OrderBy["BlockNumber"] = "blockNumber";
    TextChanged_OrderBy["TransactionId"] = "transactionID";
    TextChanged_OrderBy["Key"] = "key";
})(TextChanged_OrderBy = exports.TextChanged_OrderBy || (exports.TextChanged_OrderBy = {}));
var Transfer_OrderBy;
(function (Transfer_OrderBy) {
    Transfer_OrderBy["Id"] = "id";
    Transfer_OrderBy["Domain"] = "domain";
    Transfer_OrderBy["BlockNumber"] = "blockNumber";
    Transfer_OrderBy["TransactionId"] = "transactionID";
    Transfer_OrderBy["Owner"] = "owner";
})(Transfer_OrderBy = exports.Transfer_OrderBy || (exports.Transfer_OrderBy = {}));
var _SubgraphErrorPolicy_;
(function (_SubgraphErrorPolicy_) {
    /** Data will be returned even if the subgraph has indexing errors */
    _SubgraphErrorPolicy_["Allow"] = "allow";
    /** If the subgraph has indexing errors, data will be omitted. The default. */
    _SubgraphErrorPolicy_["Deny"] = "deny";
})(_SubgraphErrorPolicy_ = exports._SubgraphErrorPolicy_ || (exports._SubgraphErrorPolicy_ = {}));
