// @zoralabs/nft-components/dist/content-components/Image.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Image = exports.ImageRenderer = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const MediaLoader_1 = require("./MediaLoader");
const RendererConfig_1 = require("./RendererConfig");
exports.ImageRenderer = react_1.forwardRef((requestProps, ref) => {
    var _a, _b;
    const { getStyles, theme, request } = requestProps;
    const { props, loading, error } = MediaLoader_1.useMediaObjectProps({
        uri: ((_a = request.media.content) === null || _a === void 0 ? void 0 : _a.uri) || ((_b = request.media.image) === null || _b === void 0 ? void 0 : _b.uri),
        preferredIPFSGateway: theme.preferredIPFSGateway,
        ...requestProps,
    });
    return (jsx_runtime_1.jsx(MediaLoader_1.MediaLoader, Object.assign({ getStyles: getStyles, loading: loading, error: error }, { children: jsx_runtime_1.jsx("img", Object.assign({ ref: ref }, props), void 0) }), void 0));
});
exports.Image = {
    getRenderingPreference: (request) => {
        var _a, _b, _c, _d;
        if (request.media.image) {
            // Make low priority only if full screen, move to normal priority if preview
            if (request.renderingContext === "FULL" && request.media.animation) {
                return RendererConfig_1.RenderingPreference.LOW;
            }
            return RendererConfig_1.RenderingPreference.NORMAL;
        }
        if ((_b = (_a = request.media.content) === null || _a === void 0 ? void 0 : _a.type) === null || _b === void 0 ? void 0 : _b.startsWith("image/")) {
            return RendererConfig_1.RenderingPreference.NORMAL;
        }
        if ((_d = (_c = request.media.content) === null || _c === void 0 ? void 0 : _c.type) === null || _d === void 0 ? void 0 : _d.startsWith("application/svg+xml")) {
            return RendererConfig_1.RenderingPreference.NORMAL;
        }
        return RendererConfig_1.RenderingPreference.INVALID;
    },
    render: exports.ImageRenderer,
};
