// @web3-ui/core/dist/web3-ui-core.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var react = require('@chakra-ui/react');
var hooks = require('@web3-ui/hooks');
var React = require('react');
var components = require('@web3-ui/components');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

var ConnectWallet = function ConnectWallet() {
  var _useWallet = hooks.useWallet(),
      connectWallet = _useWallet.connectWallet,
      connection = _useWallet.connection,
      connected = _useWallet.connected,
      disconnectWallet = _useWallet.disconnectWallet;

  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(react.Button, {
    onClick: connected ? disconnectWallet : connectWallet
  }, connected ? connection.ens || connection.userAddress : 'Connect wallet'));
};

var Provider = function Provider(_ref) {
  var network = _ref.network,
      children = _ref.children,
      rpcUrl = _ref.rpcUrl;
  return /*#__PURE__*/React__default["default"].createElement(components.Provider, null, /*#__PURE__*/React__default["default"].createElement(hooks.Provider, {
    network: network,
    rpcUrl: rpcUrl
  }, children));
};

Object.defineProperty(exports, 'CHAIN_ID_TO_NETWORK', {
  enumerable: true,
  get: function () { return hooks.CHAIN_ID_TO_NETWORK; }
});
Object.defineProperty(exports, 'NETWORKS', {
  enumerable: true,
  get: function () { return hooks.NETWORKS; }
});
Object.defineProperty(exports, 'Web3Context', {
  enumerable: true,
  get: function () { return hooks.Web3Context; }
});
Object.defineProperty(exports, 'useTokenBalance', {
  enumerable: true,
  get: function () { return hooks.useTokenBalance; }
});
Object.defineProperty(exports, 'useTransaction', {
  enumerable: true,
  get: function () { return hooks.useTransaction; }
});
Object.defineProperty(exports, 'useWallet', {
  enumerable: true,
  get: function () { return hooks.useWallet; }
});
Object.defineProperty(exports, 'useWriteContract', {
  enumerable: true,
  get: function () { return hooks.useWriteContract; }
});
exports.ConnectWallet = ConnectWallet;
exports.Provider = Provider;
