// @zoralabs/nft-hooks/dist/fetcher/ZoraIndexerTransformers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformNFTIndexerResponse = void 0;
const tslib_1 = require("tslib");
const big_js_1 = tslib_1.__importDefault(require("big.js"));
const TransformFetchResults_1 = require("./TransformFetchResults");
function transformNFTIndexerResponse(data, auction, currency) {
    var _a, _b, _c, _d, _e;
    return {
        nft: {
            tokenId: data.tokenId.toString(),
            contract: {
                address: data.address,
                name: (_b = (_a = data.tokenContract) === null || _a === void 0 ? void 0 : _a.name) === null || _b === void 0 ? void 0 : _b.toString(),
                symbol: (_d = (_c = data.tokenContract) === null || _c === void 0 ? void 0 : _c.symbol) === null || _d === void 0 ? void 0 : _d.toString(),
            },
            owner: data.owner,
            creator: data.minter || undefined,
            metadataURI: data.tokenURI || '',
        },
        zoraNFT: data.media
            ? {
                // TODO(iain): make properly optional
                createdAtTimestamp: ((_e = data.mintTransferEvent) === null || _e === void 0 ? void 0 : _e.blockTimestamp) || 0,
                // TODO(iain): make properly optional
                contentURI: data.media.contentURI || '',
                contentHash: data.media.contentHash,
                // TODO(iain): make properly optional
                metadataURI: data.media.metadataURI || '',
                metadataHash: data.media.metadataHash,
                ownerBidShare: data.media.ownerBidShare,
                ownerBidSharePercentage: data.media.ownerBidShare
                    ? new big_js_1.default(data.media.ownerBidShare).div(new big_js_1.default(10).pow(18)).toNumber()
                    : 0,
                creatorBidShare: data.media.creatorBidShare,
                creatorBidSharePercentage: data.media.creatorBidShare
                    ? new big_js_1.default(data.media.creatorBidShare).div(new big_js_1.default(10).pow(18)).toNumber()
                    : 0,
            }
            : undefined,
        zoraIndexerResponse: data,
        pricing: TransformFetchResults_1.addAuctionInformation({
            reserve: TransformFetchResults_1.auctionDataToPricing(auction),
        }, currency),
    };
}
exports.transformNFTIndexerResponse = transformNFTIndexerResponse;
